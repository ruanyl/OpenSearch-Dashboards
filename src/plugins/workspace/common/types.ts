/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export type WorkspacePermissionItem<T extends string> = {
  modes: T[];
} & ({ type: 'user'; userId: string } | { type: 'group'; group: string });
