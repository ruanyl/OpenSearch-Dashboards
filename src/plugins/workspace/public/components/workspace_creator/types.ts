/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { App, WorkspacePermissionMode } from '../../../../../core/public';

export interface WorkspaceFeature extends Pick<App, 'dependencies'> {
  id: string;
  name: string;
}

export interface WorkspaceFeatureGroup {
  name: string;
  features: WorkspaceFeature[];
}
export interface WorkspaceFormSubmitData {
  name: string;
  description?: string;
  features?: string[];
  color?: string;
  icon?: string;
  defaultVISTheme?: string;
  permissions: WorkspacePermissionSetting[];
}

export interface WorkspaceFormData extends WorkspaceFormSubmitData {
  id: string;
  reserved?: boolean;
}

export type WorkspaceFormErrors = Omit<
  { [key in keyof WorkspaceFormData]?: string },
  'permissions'
> & {
  userPermissions?: string[];
  groupPermissions?: string[];
};

export enum WorkspacePermissionItemType {
  User = 'user',
  Group = 'group',
}

export interface TypelessPermissionSetting {
  id: string;
  modes: WorkspacePermissionMode[];
}

interface UserPermissionSetting {
  type: WorkspacePermissionItemType.User;
  userId: string;
  modes: WorkspacePermissionMode[];
}

interface GroupPermissionSetting {
  type: WorkspacePermissionItemType.Group;
  group: string;
  modes: WorkspacePermissionMode[];
}

export type WorkspacePermissionSetting = UserPermissionSetting | GroupPermissionSetting;

// when editing, attributes could be undefined in workspace form
export type WorkspaceFormEditingData = Partial<
  Omit<WorkspaceFormSubmitData, 'permissions'> & {
    userPermissions: Array<Partial<TypelessPermissionSetting>>;
    groupPermissions: Array<Partial<TypelessPermissionSetting>>;
  }
>;

export type PermissionEditingData = Array<Partial<TypelessPermissionSetting>>;
