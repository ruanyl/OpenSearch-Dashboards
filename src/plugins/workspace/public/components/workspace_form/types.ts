/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import type { WorkspacePermissionItemType, WorkspaceOperationType } from './constants';
import type { WorkspacePermissionMode } from '../../../common/constants';
import type { App, ApplicationStart } from '../../../../../core/public';

export type WorkspacePermissionSetting =
  | { type: WorkspacePermissionItemType.User; userId: string; modes: WorkspacePermissionMode[] }
  | { type: WorkspacePermissionItemType.Group; group: string; modes: WorkspacePermissionMode[] };

export interface WorkspaceFormSubmitData {
  name: string;
  description?: string;
  features?: string[];
  color?: string;
  icon?: string;
  defaultVISTheme?: string;
  permissions: WorkspacePermissionSetting[];
}

export interface WorkspaceFormData extends WorkspaceFormSubmitData {
  id: string;
  reserved?: boolean;
}

export interface WorkspaceFeature {
  id: string;
  name: string;
}

export interface WorkspaceFeatureGroup {
  name: string;
  features: WorkspaceFeature[];
}

export type WorkspaceFormErrors = Omit<
  { [key in keyof WorkspaceFormData]?: string },
  'permissions'
> & {
  permissions?: string[];
};

export interface WorkspaceFormProps {
  application: ApplicationStart;
  onSubmit?: (formData: WorkspaceFormSubmitData) => void;
  defaultValues?: WorkspaceFormData;
  operationType?: WorkspaceOperationType;
  permissionEnabled?: boolean;
  permissionLastAdminItemDeletable?: boolean;
}
