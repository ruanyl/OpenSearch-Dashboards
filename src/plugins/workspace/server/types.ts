/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  Logger,
  OpenSearchDashboardsRequest,
  RequestHandlerContext,
  SavedObjectsFindResponse,
  CoreSetup,
  WorkspacePermissionMode,
  Permissions,
  WorkspaceAttribute,
} from '../../../core/server';

export interface WorkspaceAttributeWithPermission extends WorkspaceAttribute {
  permissions: Permissions;
}

export interface WorkspaceFindOptions {
  page?: number;
  perPage?: number;
  search?: string;
  searchFields?: string[];
  sortField?: string;
  sortOrder?: string;
  permissionModes?: WorkspacePermissionMode[];
}

export interface IRequestDetail {
  request: OpenSearchDashboardsRequest;
  context: RequestHandlerContext;
  logger: Logger;
}

export interface IWorkspaceDBImpl {
  setup(dep: CoreSetup): Promise<IResponse<boolean>>;
  create(
    requestDetail: IRequestDetail,
    payload: Omit<WorkspaceAttributeWithPermission, 'id'>
  ): Promise<IResponse<{ id: WorkspaceAttribute['id'] }>>;
  list(
    requestDetail: IRequestDetail,
    options: WorkspaceFindOptions
  ): Promise<
    IResponse<
      {
        workspaces: WorkspaceAttributeWithPermission[];
      } & Pick<SavedObjectsFindResponse, 'page' | 'per_page' | 'total'>
    >
  >;
  get(
    requestDetail: IRequestDetail,
    id: string
  ): Promise<IResponse<WorkspaceAttributeWithPermission>>;
  update(
    requestDetail: IRequestDetail,
    id: string,
    payload: Omit<WorkspaceAttributeWithPermission, 'id'>
  ): Promise<IResponse<boolean>>;
  delete(requestDetail: IRequestDetail, id: string): Promise<IResponse<boolean>>;
  destroy(): Promise<IResponse<boolean>>;
}

export type IResponse<T> =
  | {
      result: T;
      success: true;
    }
  | {
      success: false;
      error?: string;
    };

export type WorkspaceRoutePermissionItem = {
  modes: Array<
    | WorkspacePermissionMode.LibraryRead
    | WorkspacePermissionMode.LibraryWrite
    | WorkspacePermissionMode.Management
  >;
} & ({ type: 'user'; userId: string } | { type: 'group'; group: string });
