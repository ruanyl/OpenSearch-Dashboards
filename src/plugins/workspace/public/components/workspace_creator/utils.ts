/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { i18n } from '@osd/i18n';
import {
  WorkspacePermissionMode,
  OptionIdToWorkspacePermissionModesMap,
  PermissionModeId,
} from '../../../../../core/public';
import {
  PermissionEditingData,
  WorkspaceFormEditingData,
  WorkspaceFormErrors,
  WorkspacePermissionSetting,
  WorkspacePermissionItemType,
  WorkspaceFormData,
  TypelessPermissionSetting,
} from './types';

export const isValidWorkspacePermissionSetting = (
  setting: Partial<WorkspacePermissionSetting>
): setting is WorkspacePermissionSetting =>
  !!setting.modes &&
  setting.modes.length > 0 &&
  ((setting.type === WorkspacePermissionItemType.User && !!setting.userId) ||
    (setting.type === WorkspacePermissionItemType.Group && !!setting.group));

export const getErrorsCount = (formErrors: WorkspaceFormErrors) => {
  let errorsCount = 0;
  if (formErrors.name) {
    errorsCount += 1;
  }
  if (formErrors.description) {
    errorsCount += 1;
  }
  if (formErrors.userPermissions) {
    errorsCount += formErrors.userPermissions.filter((permission) => !!permission).length;
  }
  if (formErrors.groupPermissions) {
    errorsCount += formErrors.groupPermissions.filter((permission) => !!permission).length;
  }
  return errorsCount;
};

export const getUserAndGroupPermissions = (
  permissions: WorkspacePermissionSetting[]
): TypelessPermissionSetting[][] => {
  const userPermissions: TypelessPermissionSetting[] = [];
  const groupPermissions: TypelessPermissionSetting[] = [];
  for (const permission of permissions) {
    if (permission.type === WorkspacePermissionItemType.User) {
      userPermissions.push({ id: permission.userId, modes: permission.modes });
    }
    if (permission.type === WorkspacePermissionItemType.Group) {
      groupPermissions.push({ id: permission.group, modes: permission.modes });
    }
  }
  return [userPermissions, groupPermissions];
};

const getUnsavedPermissionsCount = (
  initialPermissions: PermissionEditingData,
  currentPermissions: PermissionEditingData
) => {
  // for user or group permissions, unsaved changes is the sum of
  // # deleted permissions, # added permissions and # edited permissions
  let addedPermissions = 0;
  let editedPermissions = 0;
  const initialPermissionMap = new Map<string, WorkspacePermissionMode[]>();
  for (const permission of initialPermissions) {
    if (permission.id) {
      initialPermissionMap.set(permission.id, permission.modes ?? []);
    }
  }

  for (const permission of currentPermissions) {
    if (!permission.id) {
      addedPermissions += 1; // added permissions
    } else {
      const permissionModes = initialPermissionMap.get(permission.id);
      if (!permissionModes) {
        addedPermissions += 1;
      } else if (
        getPermissionModeId(permissionModes) !== getPermissionModeId(permission.modes ?? [])
      ) {
        editedPermissions += 1; // added or edited permissions
      }
    }
  }

  // currentPermissions.length = initialPermissions.length + # added permissions - # deleted permissions
  const deletedPermissions =
    addedPermissions + initialPermissions.length - currentPermissions.length;

  return addedPermissions + editedPermissions + deletedPermissions;
};

export const getUnsavedChangesCount = (
  initialFormData: WorkspaceFormData,
  currentFormData: WorkspaceFormEditingData
) => {
  let unsavedChangesCount = 0;
  if (initialFormData.name !== currentFormData.name) {
    unsavedChangesCount += 1;
  }
  // initial and current description could be undefined
  const initialDescription = initialFormData.description ?? '';
  const currentDescription = currentFormData.description ?? '';
  if (initialDescription !== currentDescription) {
    unsavedChangesCount += 1;
  }
  if (initialFormData.color !== currentFormData.color) {
    unsavedChangesCount += 1;
  }
  if (initialFormData.icon !== currentFormData.icon) {
    unsavedChangesCount += 1;
  }
  if (initialFormData.defaultVISTheme !== currentFormData.defaultVISTheme) {
    unsavedChangesCount += 1;
  }
  const featureIntersectionCount = (
    initialFormData.features?.filter((feature) => currentFormData.features?.includes(feature)) ?? []
  ).length;
  // for features, unsaved changes is the sum of # deleted features and # added features
  unsavedChangesCount += (initialFormData.features?.length ?? 0) - featureIntersectionCount;
  unsavedChangesCount += (currentFormData.features?.length ?? 0) - featureIntersectionCount;
  // for permissions, unsaved changes is the sum of # unsaved user permissions and # unsaved group permissions
  const [initialUserPermissions, initialGroupPermissions] = getUserAndGroupPermissions(
    initialFormData.permissions ?? []
  );
  unsavedChangesCount += getUnsavedPermissionsCount(
    initialUserPermissions,
    currentFormData.userPermissions ?? []
  );
  unsavedChangesCount += getUnsavedPermissionsCount(
    initialGroupPermissions,
    currentFormData.groupPermissions ?? []
  );
  return unsavedChangesCount;
};

// default permission mode is read
export const getPermissionModeId = (modes: WorkspacePermissionMode[]) => {
  for (const key in OptionIdToWorkspacePermissionModesMap) {
    if (OptionIdToWorkspacePermissionModesMap[key].every((mode) => modes?.includes(mode))) {
      return key;
    }
  }
  return PermissionModeId.Read;
};

export const getPermissionErrors = (permissions: Array<Partial<TypelessPermissionSetting>>) => {
  const permissionErrors: string[] = new Array(permissions.length);
  for (let i = 0; i < permissions.length; i++) {
    const permission = permissions[i];
    if (isValidWorkspacePermissionSetting(permission)) {
      continue;
    }
    if (!permission.id) {
      permissionErrors[i] = i18n.translate('workspace.form.permission.invalidate.id', {
        defaultMessage: 'Invalid id',
      });
      continue;
    }
    if (!permission.modes || permission.modes.length === 0) {
      permissionErrors[i] = i18n.translate('workspace.form.permission.invalidate.modes', {
        defaultMessage: 'Invalid permission modes',
      });
      continue; // this line is need for more conditions
    }
  }
  return permissionErrors;
};

export const formatPermissions = (
  userPermissions: TypelessPermissionSetting[],
  groupPermissions: TypelessPermissionSetting[]
): WorkspacePermissionSetting[] => {
  const permissions: WorkspacePermissionSetting[] = [];
  for (const permission of userPermissions) {
    permissions.push({
      userId: permission.id,
      modes: permission.modes,
      type: WorkspacePermissionItemType.User,
    });
  }
  for (const permission of groupPermissions) {
    permissions.push({
      group: permission.id,
      modes: permission.modes,
      type: WorkspacePermissionItemType.Group,
    });
  }
  return permissions;
};
