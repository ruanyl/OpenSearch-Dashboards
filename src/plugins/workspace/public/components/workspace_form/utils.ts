/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { WorkspacePermissionMode, DEFAULT_CHECKED_FEATURES_IDS } from '../../../common/constants';

import {
  WorkspaceFeature,
  WorkspaceFeatureGroup,
  WorkspacePermissionSetting,
  WorkspaceFormErrors,
} from './types';
import {
  WorkspacePermissionItemType,
  optionIdToWorkspacePermissionModesMap,
  PermissionModeId,
} from './constants';

export const isWorkspaceFeatureGroup = (
  featureOrGroup: WorkspaceFeature | WorkspaceFeatureGroup
): featureOrGroup is WorkspaceFeatureGroup => 'features' in featureOrGroup;

export const isValidWorkspacePermissionSetting = (
  setting: Partial<WorkspacePermissionSetting>
): setting is WorkspacePermissionSetting =>
  !!setting.modes &&
  setting.modes.length > 0 &&
  ((setting.type === WorkspacePermissionItemType.User && !!setting.userId) ||
    (setting.type === WorkspacePermissionItemType.Group && !!setting.group));

export const isDefaultCheckedFeatureId = (id: string) => {
  return DEFAULT_CHECKED_FEATURES_IDS.indexOf(id) > -1;
};

export const appendDefaultFeatureIds = (ids: string[]) => {
  // concat default checked ids and unique the result
  return Array.from(new Set(ids.concat(DEFAULT_CHECKED_FEATURES_IDS)));
};

export const isValidNameOrDescription = (input?: string) => {
  if (!input) {
    return true;
  }
  const regex = /^[0-9a-zA-Z()_\[\]\-\s]+$/;
  return regex.test(input);
};

export const getNumberOfErrors = (formErrors: WorkspaceFormErrors) => {
  let numberOfErrors = 0;
  if (formErrors.name) {
    numberOfErrors += 1;
  }
  if (formErrors.description) {
    numberOfErrors += 1;
  }
  if (formErrors.permissions) {
    numberOfErrors += formErrors.permissions.length;
  }
  return numberOfErrors;
};

export const isUserOrGroupPermissionSettingDuplicated = (
  permissionSettings: Array<Partial<WorkspacePermissionSetting>>,
  permissionSettingToCheck: WorkspacePermissionSetting
) =>
  permissionSettings.some(
    (permissionSetting) =>
      (permissionSettingToCheck.type === WorkspacePermissionItemType.User &&
        permissionSetting.type === WorkspacePermissionItemType.User &&
        permissionSettingToCheck.userId === permissionSetting.userId) ||
      (permissionSettingToCheck.type === WorkspacePermissionItemType.Group &&
        permissionSetting.type === WorkspacePermissionItemType.Group &&
        permissionSettingToCheck.group === permissionSetting.group)
  );

export const generateWorkspacePermissionItemKey = (
  item: Partial<WorkspacePermissionSetting>,
  index?: number
) =>
  [
    ...(item.type ?? []),
    ...(item.type === WorkspacePermissionItemType.User ? [item.userId] : []),
    ...(item.type === WorkspacePermissionItemType.Group ? [item.group] : []),
    ...(item.modes ?? []),
    index,
  ].join('-');

// default permission mode is read
export const getPermissionModeId = (modes: WorkspacePermissionMode[]) => {
  for (const key in optionIdToWorkspacePermissionModesMap) {
    if (optionIdToWorkspacePermissionModesMap[key].every((mode) => modes?.includes(mode))) {
      return key;
    }
  }
  return PermissionModeId.Read;
};
