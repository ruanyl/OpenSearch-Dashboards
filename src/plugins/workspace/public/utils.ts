/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  AppCategory,
  PublicAppInfo,
  AppNavLinkStatus,
  DEFAULT_APP_CATEGORIES,
} from '../../../core/public';

/**
 * Checks if a given feature matches the provided feature configuration.
 *
 * Rules:
 * 1. `*` matches any feature.
 * 2. Config starts with `@` matches category, for example, @management matches any feature of `management` category,
 * 3. To match a specific feature, use the feature id, such as `discover`,
 * 4. To exclude a feature or category, prepend with `!`, e.g., `!discover` or `!@management`.
 * 5. The order of featureConfig array matters. From left to right, later configs override the previous ones.
 *    For example, ['!@management', '*'] matches any feature because '*' overrides the previous setting: '!@management'.
 */
export const featureMatchesConfig = (featureConfigs: string[]) => ({
  id,
  category,
}: {
  id: string;
  category?: AppCategory;
}) => {
  let matched = false;

  for (const featureConfig of featureConfigs) {
    // '*' matches any feature
    if (featureConfig === '*') {
      matched = true;
    }

    // The config starts with `@` matches a category
    if (category && featureConfig === `@${category.id}`) {
      matched = true;
    }

    // The config matches a feature id
    if (featureConfig === id) {
      matched = true;
    }

    // If a config starts with `!`, such feature or category will be excluded
    if (featureConfig.startsWith('!')) {
      if (category && featureConfig === `!@${category.id}`) {
        matched = false;
      }

      if (featureConfig === `!${id}`) {
        matched = false;
      }
    }
  }

  return matched;
};

// Get all apps excluding management category
export const getAllExcludingManagementApps = (applications: PublicAppInfo[]): PublicAppInfo[] => {
  return applications.filter(
    ({ navLinkStatus, chromeless, category }) =>
      navLinkStatus !== AppNavLinkStatus.hidden &&
      !chromeless &&
      category?.id !== DEFAULT_APP_CATEGORIES.management.id
  );
};

export const getSelectedFeatureQuantities = (
  featuresConfig: string[],
  applications: PublicAppInfo[]
) => {
  const visibleApplications = getAllExcludingManagementApps(applications);
  const featureFilter = featureMatchesConfig(featuresConfig);
  const selectedApplications = visibleApplications.filter((app) => featureFilter(app));
  return {
    total: visibleApplications.length,
    selected: selectedApplications.length,
  };
};
