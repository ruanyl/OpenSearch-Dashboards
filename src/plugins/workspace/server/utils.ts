/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import crypto from 'crypto';
import {
  AuthStatus,
  HttpAuth,
  OpenSearchDashboardsRequest,
  Principals,
  PrincipalType,
} from '../../../core/server';
import { AuthInfo } from './types';

/**
 * Generate URL friendly random ID
 */
export const generateRandomId = (size: number) => {
  return crypto.randomBytes(size).toString('base64url').slice(0, size);
};

export const getPrincipalsFromRequest = (
  request: OpenSearchDashboardsRequest,
  auth?: HttpAuth
): Principals => {
  const payload: Principals = {};
  const authInfoResp = auth?.get(request);
  if (authInfoResp?.status === AuthStatus.unknown) {
    /**
     * Login user have access to all the workspaces when no authentication is presented.
     */
    return payload;
  }

  if (authInfoResp?.status === AuthStatus.authenticated) {
    const authInfo = authInfoResp?.state as AuthInfo | null;
    if (authInfo?.backend_roles) {
      payload[PrincipalType.Groups] = authInfo.backend_roles;
    }
    if (authInfo?.user_name) {
      payload[PrincipalType.Users] = [authInfo.user_name];
    }
    return payload;
  }

  if (authInfoResp?.status === AuthStatus.unauthenticated) {
    throw new Error('NOT_AUTHORIZED');
  }

  throw new Error('UNEXPECTED_AUTHORIZATION_STATUS');
};
