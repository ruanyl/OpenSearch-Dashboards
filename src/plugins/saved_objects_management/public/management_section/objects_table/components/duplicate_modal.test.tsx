/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  httpServiceMock,
  notificationServiceMock,
  workspacesServiceMock,
} from '../../../../../../core/public/mocks';
import {
  DuplicateMode,
  ShowDuplicateModalProps,
  SavedObjectsDuplicateModal,
} from './duplicate_modal';
import { SavedObjectWithMetadata } from '../../../types';
import { shallowWithI18nProvider } from 'test_utils/enzyme_helpers';
import React from 'react';
import { WorkspaceObject } from 'src/core/types';
import { WorkspaceOption } from './utils';

interface Props extends ShowDuplicateModalProps {
  onClose: () => void;
}

describe('DuplicateModal', () => {
  let selectedModeProps: Props;
  let allModeProps: Props;
  let http: ReturnType<typeof httpServiceMock.createStartContract>;
  let notifications: ReturnType<typeof notificationServiceMock.createStartContract>;
  let workspaces: ReturnType<typeof workspacesServiceMock.createStartContract>;
  const selectedSavedObjects: SavedObjectWithMetadata[] = [
    {
      id: '1',
      type: 'dashboard',
      workspaces: ['workspace1'],
      attributes: {},
      references: [],
      meta: {
        title: 'Dashboard_1',
        icon: 'dashboardApp',
      },
    },
    {
      id: '2',
      type: 'visualization',
      workspaces: ['workspace2'],
      attributes: {},
      references: [],
      meta: {
        title: 'Visualization',
        icon: 'visualizationApp',
      },
    },
    {
      id: '3',
      type: 'dashboard',
      workspaces: ['workspace2'],
      attributes: {},
      references: [],
      meta: {
        title: 'Dashboard_2',
        icon: 'dashboardApp',
      },
    },
  ];
  const workspaceList: WorkspaceObject[] = [
    {
      id: 'workspace1',
      name: 'foo',
    },
    {
      id: 'workspace2',
      name: 'bar',
    },
  ];
  beforeEach(() => {
    http = httpServiceMock.createStartContract();
    notifications = notificationServiceMock.createStartContract();
    workspaces = workspacesServiceMock.createStartContract();

    selectedModeProps = {
      onDuplicate: jest.fn(),
      onClose: jest.fn(),
      http,
      workspaces,
      duplicateMode: DuplicateMode.Selected,
      notifications,
      selectedSavedObjects,
    };

    allModeProps = {
      ...selectedModeProps,
      duplicateMode: DuplicateMode.All,
      selectedSavedObjects,
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render normally', async () => {
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    // Ensure all promises resolve
    await new Promise((resolve) => process.nextTick(resolve));
    // Ensure the state changes are reflected
    component.update();
    expect(component).toMatchSnapshot();
  });

  it('should show all target workspace options when not in any workspace', async () => {
    workspaces.workspaceList$.next(workspaceList);
    workspaces.currentWorkspaceId$.next('');
    workspaces.currentWorkspace$.next(null);
    selectedModeProps = { ...selectedModeProps, workspaces };
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    await new Promise((resolve) => process.nextTick(resolve));
    component.update();
    const options = component.find('EuiComboBox').prop('options') as WorkspaceOption[];
    expect(options.length).toEqual(2);
    expect(options[0].label).toEqual('foo');
    expect(options[1].label).toEqual('bar');
  });

  it('should display the suffix (current) in target workspace options when in workspace1', async () => {
    workspaces.workspaceList$.next(workspaceList);
    workspaces.currentWorkspaceId$.next('workspace1');
    workspaces.currentWorkspace$.next({
      id: 'workspace1',
      name: 'foo',
    });
    selectedModeProps = { ...selectedModeProps, workspaces };
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    await new Promise((resolve) => process.nextTick(resolve));
    component.update();
    const options = component.find('EuiComboBox').prop('options') as WorkspaceOption[];
    expect(options.length).toEqual(2);
    expect(options[0].label).toEqual('foo (current)');
    expect(options[1].label).toEqual('bar');
  });

  it('should only show saved objects belong to workspace1 when target workspace is workspace2', async () => {
    workspaces.workspaceList$.next(workspaceList);
    workspaces.currentWorkspaceId$.next('');
    workspaces.currentWorkspace$.next(null);
    selectedModeProps = { ...selectedModeProps, workspaces };
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    const selectedObjects = component
      .find('EuiInMemoryTable')
      .prop('items') as SavedObjectWithMetadata[];
    expect(selectedObjects.length).toEqual(3);
    const comboBox = component.find('EuiComboBox');
    comboBox.simulate('change', [{ label: 'bar', key: 'workspace2', value: workspaceList[1] }]);
    component.update();

    const targetWorkspaceOption = component.state('targetWorkspaceOption') as WorkspaceOption[];
    expect(targetWorkspaceOption.length).toEqual(1);
    expect(targetWorkspaceOption[0].key).toEqual('workspace2');

    const includedSelectedObjects = component
      .find('EuiInMemoryTable')
      .prop('items') as SavedObjectWithMetadata[];
    expect(includedSelectedObjects.length).toEqual(1);
    expect(includedSelectedObjects[0].workspaces).toEqual(['workspace1']);
    expect(includedSelectedObjects[0].id).toEqual('1');

    expect(component.find('EuiCallOut').prop('aria-disabled')).toEqual(false);
  });

  it('should show saved objects type when duplicate mode is all', async () => {
    workspaces.workspaceList$.next(workspaceList);
    workspaces.currentWorkspaceId$.next('');
    workspaces.currentWorkspace$.next(null);
    allModeProps = { ...allModeProps, workspaces };
    const component = shallowWithI18nProvider(<SavedObjectsDuplicateModal {...allModeProps} />);
    const savedObjectTypeInfoMap = component.state('savedObjectTypeInfoMap') as Map<
      string,
      [number, boolean]
    >;
    expect(savedObjectTypeInfoMap.get('dashboard')).toEqual([2, true]);
  });

  it('should call onClose function when cancle button is clicked', () => {
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    component.find('[data-test-subj="duplicateCancelButton"]').simulate('click');
    expect(selectedModeProps.onClose).toHaveBeenCalled();
  });

  it('should call onDuplicate function when confirm button is clicked', () => {
    workspaces.workspaceList$.next(workspaceList);
    workspaces.currentWorkspaceId$.next('');
    workspaces.currentWorkspace$.next(null);
    selectedModeProps = { ...selectedModeProps, workspaces };
    const component = shallowWithI18nProvider(
      <SavedObjectsDuplicateModal {...selectedModeProps} />
    );
    const comboBox = component.find('EuiComboBox');
    comboBox.simulate('change', [{ label: 'bar', key: 'workspace2', value: workspaceList[1] }]);
    const confirmButton = component.find('[data-test-subj="duplicateConfirmButton"]');
    expect(confirmButton.prop('isLoading')).toBe(false);
    expect(confirmButton.prop('disabled')).toBe(false);
    confirmButton.simulate('click');
    expect(selectedModeProps.onDuplicate).toHaveBeenCalled();
  });
});
