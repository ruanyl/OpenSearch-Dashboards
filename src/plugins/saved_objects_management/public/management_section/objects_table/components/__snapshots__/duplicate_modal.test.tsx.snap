// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DuplicateModal should Unmount normally 1`] = `""`;

exports[`DuplicateModal should render normally 1`] = `
<EuiModal
  className="savedObjectsModal"
  data-test-subj="savedObjectsDuplicateModal"
  onClose={[MockFunction]}
>
  <EuiModalHeader>
    <EuiModalHeaderTitle>
      <FormattedMessage
        defaultMessage="Duplicate {duplicateMode, select, all {all objects} other {{objectCount, plural, =1 {{objectName}} other {# objects}}}}?"
        id="savedObjectsManagement.objectsTable.duplicateModal.title"
        values={
          Object {
            "duplicateMode": "selected",
            "objectCount": 3,
            "objectName": "Dashboard_1",
          }
        }
      />
    </EuiModalHeaderTitle>
  </EuiModalHeader>
  <EuiModalBody>
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={true}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label="Destination workspace"
      labelType="label"
    >
      <EuiText
        color="subdued"
        size="s"
      >
        Specify a workspace where the objects will be duplicated.
      </EuiText>
      <EuiSpacer
        size="s"
      />
      <EuiComboBox
        async={false}
        compressed={false}
        fullWidth={false}
        isClearable={false}
        isInvalid={true}
        onChange={[Function]}
        options={Array []}
        selectedOptions={Array []}
        singleSelection={
          Object {
            "asPlainText": true,
          }
        }
        sortMatchesBy="none"
      />
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <EuiFormRow
      describedByIds={Array []}
      display="row"
      fullWidth={true}
      hasChildLabel={true}
      hasEmptyLabelSpace={false}
      label="Related Objects"
      labelType="label"
    >
      <EuiText
        color="subdued"
        size="s"
      >
        We recommended duplicating related objects to ensure your duplicated objects will continue to function.
      </EuiText>
      <EuiSpacer
        size="s"
      />
      <EuiCheckbox
        checked={true}
        compressed={false}
        disabled={false}
        id="includeReferencesDeep"
        indeterminate={false}
        label="Duplicate related objects"
        onChange={[Function]}
      />
    </EuiFormRow>
    <EuiSpacer
      size="m"
    />
    <p>
      <FormattedMessage
        defaultMessage="The following saved objects will be copied:"
        id="savedObjectsManagement.objectsTable.duplicateModal.tableTitle"
        values={Object {}}
      />
    </p>
    <EuiSpacer
      size="m"
    />
    <EuiInMemoryTable
      columns={
        Array [
          Object {
            "field": "type",
            "name": "Type",
            "render": [Function],
            "width": "50px",
          },
          Object {
            "field": "id",
            "name": "Id",
          },
          Object {
            "field": "meta.title",
            "name": "Title",
          },
        ]
      }
      items={
        Array [
          Object {
            "attributes": Object {},
            "id": "1",
            "meta": Object {
              "icon": "dashboardApp",
              "title": "Dashboard_1",
            },
            "references": Array [],
            "type": "dashboard",
            "workspaces": Array [
              "workspace1",
            ],
          },
          Object {
            "attributes": Object {},
            "id": "2",
            "meta": Object {
              "icon": "visualizationApp",
              "title": "Visualization",
            },
            "references": Array [],
            "type": "visualization",
            "workspaces": Array [
              "workspace2",
            ],
          },
          Object {
            "attributes": Object {},
            "id": "3",
            "meta": Object {
              "icon": "dashboardApp",
              "title": "Dashboard_2",
            },
            "references": Array [],
            "type": "dashboard",
            "workspaces": Array [
              "workspace2",
            ],
          },
        ]
      }
      pagination={true}
      responsive={true}
      sorting={false}
      tableLayout="fixed"
    />
  </EuiModalBody>
  <EuiModalFooter>
    <EuiButtonEmpty
      data-test-subj="duplicateCancelButton"
      onClick={[MockFunction]}
    >
      <FormattedMessage
        defaultMessage="Cancel"
        id="savedObjectsManagement.objectsTable.duplicateModal.duplicateCancelButton"
        values={Object {}}
      />
    </EuiButtonEmpty>
    <EuiButton
      data-test-subj="duplicateConfirmButton"
      disabled={true}
      fill={true}
      isLoading={false}
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Duplicate{duplicateMode, select, all {({objectCount})} other {}}"
        id="savedObjectsManagement.objectsTable.duplicateModal.confirmButtonLabel"
        values={
          Object {
            "duplicateMode": "selected",
            "objectCount": 3,
          }
        }
      />
    </EuiButton>
  </EuiModalFooter>
</EuiModal>
`;
